### ISSUE TITLE
Rabbit MQ management UI is not accessing when running with nixos 17.09 + docker 17.09

### ISSUE THREAD

**Issue by kuznero** [NixOS/nixpkgs#30289]
## Issue description

When running `rabbitmq:3.6.12-management` docker image in single-noded swarm cluster, it is impossible to connect to management UI through exposed port.

This behavior started to show up on latest nixos 17.09 with latest docker 17.09. The following combinations were also tested:

* `nixos 17.09` - `docker 17.09` - **not working**
* `nixos 17.09` - `docker 17.06` - **not working**
* `nixos 17.09` - `docker 17.03` (from `nixos-17.03` channel) - **not working**
* `nixos 17.03` - `docker 17.03` - **works as expected**

### Steps to reproduce

```bash
$ docker swarm init
$ docker network create ci -d overlay
$ docker service create --name rabbit --network ci -p 15672:15672 rabbitmq:3.6.12-management
```

Now, if to open [http://127.0.0.1:15672/](http://127.0.0.1:15672/) in a browser it will never show anything and spinner will go crazy forever.

## Technical details

* System: 17.09.1547.2231575a7d (Hummingbird)
* Nix version: nix-env (Nix) 1.11.15
* Nixpkgs version: 17.09.1547.2231575a7d
* Sandboxing enabled: build-use-sandbox = true

**Comment by kuznero** [comment]
Ping @grahamc

**Comment by kuznero** [comment]
Have also test the scenario of a freshly installed machine with nixos 17.09. Exactly same behavior.

**Comment by kuznero** [comment]
Here is the result of `journalctl -xe | grep dockerd`:

```
level=info msg="libcontainerd: new containerd process, pid: 940"
level=info msg="[graphdriver] using prior storage driver: overlay2"
level=info msg="Graph migration to content-addressability took 0.00 seconds"
level=warning msg="Your kernel does not support cgroup cfs period"
level=warning msg="Your kernel does not support cgroup cfs quotas"
level=warning msg="Your kernel does not support cgroup rt period"
level=warning msg="Your kernel does not support cgroup rt runtime"
level=info msg="Loading containers: start."
level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
level=info msg="Loading containers: done."
level=warning msg="failed to retrieve docker-runc version: unknown output format: runc version 1.0.0-rc4+dev spec: 1.0.0"
level=info msg="Docker daemon" commit=afdb6d44a80f777069885a9ee0e0f86cf841b1bb graphdriver(s)=overlay2 version=17.09.0-ce
level=info msg="Daemon has completed initialization"
level=info msg="API listen on 0.0.0.0:2375"
level=info msg="API listen on /var/run/docker.sock"
level=info msg="Processing signal 'terminated'"
level=info msg="stopping containerd after receiving terminated"
```

**Comment by kuznero** [comment]
After downgrading from nixos 17.09 to nixos 17.03, I can see this summary:

![capture](https://user-images.githubusercontent.com/449813/31437574-0818b6ba-ae86-11e7-9483-d3c29f5591ab.PNG)

Please notice that `NetworkManager-dispatcher` and `systemd-hostnamed` services haven't been running in nixos 17.09, and only started when I rolled back to nixos 17.03. Can that be the reason for this weird behavior?

**Comment by ArdaXi** [comment]
I don't have a swarm up, but `docker run -p 15672:15672 rabbitmq:3.6.12-management` works fine on my `18.03.git.9824ca6 (Impala)`.

Those same 5 errors show up in my docker log so those are not causing this issue.

**Comment by kuznero** [comment]
Here is a bit more details with `netstat -nlp | grep 15672`:

```
tcp6       0      0 :::15672                :::*                    LISTEN      771/dockerd
```

And also with `netstat -a | grep 15672`:

```
tcp        0      0 localhost:53564         localhost:15672         ESTABLISHED                      
tcp6       0      0 [::]:15672              [::]:*                  LISTEN     
```

Notice, that the line with `ESTABLISHED` shows up when I try opening [http://127.0.0.1:15672](http://127.0.0.1:15672) in my browser.

**Comment by aszlig** [comment]
Disclaimer: I'm not using docker and I also don't want to run it on my machine, but here is an incomplete VM test that could be expanded to reproduce the problem (and which can also serve as a regression test once the bug has been found and fixed):

```nix
let
  dockerConfig = { pkgs, ... }: {
    virtualisation.docker.enable = true;
    virtualisation.docker.package = pkgs.docker;
    virtualisation.docker.liveRestore = false;
  };

in import <nixpkgs/nixos/tests/make-test.nix> {
  name = "docker-swarm";

  machine = { config, pkgs, ... }: let
    image = (import <nixpkgs/nixos/lib/testing.nix> {
      inherit (config.nixpkgs) system;
    }).runInMachine {
      drv = pkgs.runCommand "docker-image" {
        outputHashAlgo = "sha256";
        outputHash = "0jzlnbvc6yvrmaw0yj7bv6nlj60bnzl4178calmgcw43xgm38amw";
        outputHashMode = "recursive";
        nativeBuildInputs = [ pkgs.docker ];
      } ''
        docker pull rabbitmq:3.6.12-management
        docker save rabbitmq:3.6.12-management > "$out/image.tar"
      '';
      machine = dockerConfig;
    };

  in {
    system.build.docker-image = image;
    imports = [ dockerConfig ];
  };

  testScript = { nodes, ... }: let
    inherit (nodes.machine.config.networking) primaryIPAddress;
    inherit (nodes.machine.config.system.build) docker-image;
  in ''
    $machine->waitForUnit('multi-user.target');
    $machine->succeed(
      'docker swarm init --advertise-addr=${primaryIPAddress}',
      'docker network create ci -d overlay',
      'docker load -i ${docker-image}/image.tar',
      'docker service create --name rabbit -p 15672:15672'.
      ' rabbitmq:3.6.12-management'
    );
  '';
}
```

The issue left to solve is that `docker service create` tries to access the network:

```
machine: must succeed: docker service create --name rabbit -p 15672:15672 rabbitmq
machine# [   58.166816] dockerd[792]: time="2017-10-11T18:16:01.043841862Z" level=warning msg="failed to retrieve docker-runc version: unknown output format: runc version 1.0.0-rc4+dev
machine# [   58.169484] dockerd[792]: spec: 1.0.0
machine# [   58.171450] dockerd[792]: "
machine# [   59.784536] dockerd[792]: time="2017-10-11T18:16:02.661413928Z" level=error msg="Handler for GET /v1.32/distribution/rabbitmq:latest/json returned error: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)"
```

So I'd appreciate it if some of the docker users here have an idea on how to run `docker service create` without network access.

Another issue with this test is that the `docker-image` derivation is very likely to be not deterministic, so in order to add such a test to `<nixpkgs>` we'd need to make sure that the output is deterministic, maybe docker has a way to specify to pull an _exact_ version of a particular image?

Cc: @grahamc

**Comment by aszlig** [comment]
Cc: @offlinehacker, @abbradar, @Mic92, @bachp, @lethalman, @bjornfor, @bennofs, @NeQuissimus, @shlevy, @rasendubi, @ragnard, @tailhook, @oxij, @Fuuzetsu, @fpletz, @bosu
... aka everybody who has made changes to the docker NixOS module.

**Comment by NeQuissimus** [comment]
This should pull the same image every time:

`docker pull rabbitmq:3.6.12-management@sha256:bcdfca2f591e9d8e88d4002305b480b036e4b9a1ab26b517e853446f2d20b815`

**Comment by NeQuissimus** [comment]
I didn't have much time but this should get you further:
For some reason, the `image.tar` is "file not found"

```nix
let
  dockerConfig = { pkgs, ... }: {
    virtualisation.docker.enable = true;
    virtualisation.docker.package = pkgs.docker;
    virtualisation.docker.liveRestore = false;
  };

in import <nixpkgs/nixos/tests/make-test.nix> {
  name = "docker-swarm";

  machine = { config, pkgs, ... }: let
    image = (import <nixpkgs/nixos/lib/testing.nix> {
      inherit (config.nixpkgs) system;
    }).runInMachine {
      drv = pkgs.runCommand "docker-image" {
        outputHashAlgo = "sha256";
        outputHash = "0sjjj9z1dhilhpc8pq4154czrb79z9cm044jvn75kxcjv6v5l2m5";
        outputHashMode = "recursive";
        nativeBuildInputs = [ pkgs.docker ];
      } ''
        docker pull rabbitmq:3.6.12-management@sha256:bcdfca2f591e9d8e88d4002305b480b036e4b9a1ab26b517e853446f2d20b815
        docker save rabbitmq:3.6.12-management@sha256:bcdfca2f591e9d8e88d4002305b480b036e4b9a1ab26b517e853446f2d20b815 > "$out/image.tar"
      '';
      machine = dockerConfig;
    };

  in {
    system.build.docker-image = image;
    imports = [ dockerConfig ];
  };

  testScript = { nodes, ... }: let
    inherit (nodes.machine.config.networking) primaryIPAddress;
    inherit (nodes.machine.config.system.build) docker-image;
  in ''
    $machine->waitForUnit('multi-user.target');
    $machine->succeed(
      'docker swarm init --advertise-addr=${primaryIPAddress}',
      'docker network create ci -d overlay',
      'docker load -i ${docker-image}/image.tar',
      'docker service create --name rabbit -p 15672:15672 $(docker images | tail -1 | awk \'{print $3}\')'
    );
  '';
}
```

**Comment by aszlig** [comment]
@NeQuissimus: The reason why it says `file not found` is that `runInMachine` has been broken since a long time, but it I've fixed it in 20487112ed9cbc3fe0d5b92e0b861a7c09a33735.

Okay, that helps a lot, trying to get it working...

Btw. shouldn't `rabbitmq:3.6.12-management@sha256:bcdfca2f591e9d8e88d4002305b480b036e4b9a1ab26b517e853446f2d20b815` be enough?

**Comment by aszlig** [comment]
@NeQuissimus: Current version of the VM test (this time as a gist so we don't need to post multiple iterations of it here): https://gist.github.com/aszlig/0c845fa29645866d5f879feac3f9d194

Unfortunately, `docker service create` still has issues even with the image ID (decluttered and without timestamps):

```
[warning] Error getting v2 registry: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: connection timed out
[info] Attempting next endpoint for pull after error: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: connection timed out
[error] pulling image failed: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: connection timed out
[error] fatal task error: No such image: 046c07b70859:latest
[warning] failed to retrieve docker-runc version: unknown output format: runc version 1.0.0-rc4+dev
          spec: 1.0.0
[warning] Error getting v2 registry: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[info] Attempting next endpoint for pull after error: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[error] pulling image failed: Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
[error] fatal task error: No such image: 046c07b70859:latest
[warning] Error getting v2 registry: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: connection timed out
[info] Attempting next endpoint for pull after error: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: connection timed out
[error] pulling image failed: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io: connection timed out
[error] fatal task error: No such image: 046c07b70859:latest
[warning] failed to retrieve docker-runc version: unknown output format: runc version 1.0.0-rc4+dev
```

The main point here is that it says `No such image: 046c07b70859:latest`, but that's exactly the image ID which is reported by `docker images`. I've also tried to use other variants like `rabbitmq@sha256:bcdfca2f591e9d8e88d4002305b480b036e4b9a1ab26b517e853446f2d20b815` or just `rabbitmq` to no extent.

**Comment by NeQuissimus** [comment]
I forked to https://gist.github.com/NeQuissimus/a902ce4192f24877b8c78b1bdde7e7c2 and the test passes

**Comment by NeQuissimus** [comment]
Back to the issue at hand... @kuznero How long did you wait until you have up on the container? I find that Docker swarm takes quite a bit of time to full start up RabbitMQ (in my test up to 3 minutes)
The test I just posted above seems to verify that it does (eventually) work.

**Comment by kuznero** [comment]
@NeQuissimus I wait about 20-30 seconds typically. Container will say that it started about 6 plugins and at about that time it should be possible to open management UI in browser or simply run curl.

**Comment by NeQuissimus** [comment]
Can you just let it sit for 5 minutes and then try?
This may just be a timing thing rather than it not working at all... (which could be due to a large variety of things, including those warnings you cited)

**Comment by NeQuissimus** [comment]
Also: **Nobody close this issue until the test is in nixpkgs!**

**Comment by kuznero** [comment]
@NeQuissimus I have tried that after 5 minutes and also later. Still the same.

**Comment by NeQuissimus** [comment]
hmmmm..  from what I can tell Docker should be the same in unstable and 17.09. I usually cherry-pick it.
I assume your system has configuration similar to what is in the test?

```nix
virtualisation.docker.enable = true;
virtualisation.docker.liveRestore = false;
```

**Comment by kuznero** [comment]
Yes, @NeQuissimus - enable and no live restore. Here is what I have approximately:

```nix
docker = {
  enable = true;
  enableOnBoot = true;
  liveRestore = false;
  logDriver = "json-file";
  listenOptions = [ "/var/run/docker.sock" "0.0.0.0:2375" ];
  extraOptions = ''
    --log-opt max-size=1M --log-opt max-file=1
  '';
};
```

**Comment by NeQuissimus** [comment]
Well, that's just odd... Running the image directly definitely works. The test passes for Swarm. When I have a chance, I can try out Swarm on my machine but since I use Docker myself, I hate to mess with it too much. Good thing, I can roll back generations :)

**Comment by kuznero** [comment]
Should we try to confirm that it is or it isn't the same with let's say Ubuntu as a host? I will try to see today if I can test it.

**Comment by NeQuissimus** [comment]
That would be great. Sorry, I've been a little swamped the last couple of weeks.

**Comment by kuznero** [comment]
Took latest `ubuntu 17.04`, installed latest `docker ce 17.09`. Tried same scenario:

1. init single noded swarm node
2. create overlay network
3. start a rabbitmq service on that network exposing `15672` port
4. open browser and navigate to `http://127.0.0.1:15672`

All works! I think there is something in NixOS in the end of the day :(

**Comment by NeQuissimus** [comment]
(I am just spamming what I am doing right now)



OK, here is what I get in my journal:

```
Oct 19 07:43:23 nixus dockerd[4474]: time="2017-10-19T07:43:23.429265895-04:00" level=info msg="IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844]"
Oct 19 07:43:23 nixus dockerd[4474]: time="2017-10-19T07:43:23.449767410-04:00" level=error msg="fatal task error" error="starting container failed: subnet sandbox join failed for \"10.255.0.0/16\": error creating vxlan interface: operation not supported" module="node/agent/taskmanager" node.id=5v06hk9wid3uqbrgbu4g3kuh1 service.id=c9hqt12xcigcewuhcrvsjdqhc task.id=wdonqng0atweuptgubxey9nv1
```

My containers never come further than "Created"

```
λ docker ps -a
CONTAINER ID        IMAGE                        COMMAND                  CREATED              STATUS              PORTS               NAMES
e69aed35700b        rabbitmq:3.6.12-management   "docker-entrypoint..."   3 seconds ago        Created                                 rabbit.1.k8i6ax2e9n73ll5l188vnh6wn
b55b460a9224        rabbitmq:3.6.12-management   "docker-entrypoint..."   24 seconds ago       Created                                 rabbit.1.wdonqng0atweuptgubxey9nv1
4437845b15e6        rabbitmq:3.6.12-management   "docker-entrypoint..."   About a minute ago   Created                                 rabbit.1.v226cvupvra2k7gicea9lcuk6
bf045c8cac83        rabbitmq:3.6.12-management   "docker-entrypoint..."   About a minute ago   Created                                 rabbit.1.wr3m29940gf1hcr5meeeo0586
ed321c4daf85        rabbitmq:3.6.12-management   "docker-entrypoint..."   About a minute ago   Created                                 rabbit.1.r06y4z1irooxkf9eas7mmzhpb
```

Turns out I need the `vxlan` kernel module as per docker/machine#2753 (which is included by default but I have dynamic module loading disabled)

**Comment by NeQuissimus** [comment]
Now I see this:

```
Oct 19 08:05:06 nixus dockerd[1383]: time="2017-10-19T08:05:06-04:00" level=error msg="setting up rule failed, [-t nat -A PREROUTING -d 10.255.0.4 -p tcp --dport 15672 -j REDIRECT --to-port 15672]:  (iptables failed: iptables --wait -t nat -A PREROUTING -d 10.255.0.4 -p tcp --dport 15672 -j REDIRECT --to-port 15672: iptables: No chain/target/match by that name.
Oct 19 08:05:06 nixus dockerd[1383]:  (exit status 1))"
Oct 19 08:05:06 nixus dockerd[1383]: time="2017-10-19T08:05:06.557735348-04:00" level=error msg="Failed to add redirect rules for ep rabbit.1.y3zr2dotv7h5f7qyrjlixykt5 (42c37fb): reexec failed: exit status 6"
Oct 19 08:05:06 nixus dockerd[1383]: time="2017-10-19T08:05:06.632841070-04:00" level=warning msg="Running modprobe ip_vs failed with message: `modprobe: ERROR: could not insert 'ip_vs': Operation not permitted
Oct 19 08:05:06 nixus dockerd[1383]: insmod /run/current-system/kernel-modules/lib/modules/4.13.7-hardened/kernel/net/netfilter/ipvs/ip_vs.ko.xz`, error: exit status 1"
Oct 19 08:05:06 nixus dockerd[1383]: time="2017-10-19T08:05:06.634613982-04:00" level=error msg="Could not get ipvs family information from the kernel. It is possible that ipvs is not enabled in your kernel. Native loadbalancing will not work until this is fixed."
```

Adding `ip_vs` :)

**Comment by NeQuissimus** [comment]
Now I have this over and over:

```
Oct 19 08:24:51 nixus dockerd[1403]: time="2017-10-19T08:24:51.389484913-04:00" level=error msg="Failed to delete real server 10.255.0.4 for vip 10.255.0.3 fwmark 257 in sbox ingress (ingress
): no such process"
Oct 19 08:24:51 nixus dockerd[1403]: time="2017-10-19T08:24:51.389529167-04:00" level=error msg="Failed to delete service for vip 10.255.0.3 fwmark 257 in sbox ingress (ingress): no such proc
ess"
Oct 19 08:24:51 nixus dockerd[1403]: time="2017-10-19T08:24:51-04:00" level=error msg="setting up rule failed, [-t mangle -D OUTPUT -d 10.255.0.3/32 -j MARK --set-mark 257]:  (iptables failed
: iptables --wait -t mangle -D OUTPUT -d 10.255.0.3/32 -j MARK --set-mark 257: iptables: No chain/target/match by that name.
Oct 19 08:24:51 nixus dockerd[1403]:  (exit status 1))"
Oct 19 08:24:51 nixus dockerd[1403]: time="2017-10-19T08:24:51.446991045-04:00" level=error msg="Failed to delete firewall mark rule in sbox ingress (ingress): reexec failed: exit status 5"
Oct 19 08:24:51 nixus dockerd[1403]: time="2017-10-19T08:24:51.542001034-04:00" level=error msg="failed removing service binding for 2af4235f16876d373eddae9449189a3b35e011c8340a9ec1870db0b42a
9ff027 epRec:{rabbit.1.lfdil0zokeud3fa6u0zt0f1yp rabbit kf38e3f4mxlu6i35rzdukip46 10.0.0.2 10.0.0.3 [] [] [e23a8d00aeb4]} err:network u6de2anic31s6g51g9cnzdzl7 not found"
Oct 19 08:24:55 nixus dockerd[1403]: time="2017-10-19T08:24:55.848672427-04:00" level=info msg="No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [name
server 8.8.8.8 nameserver 8.8.4.4]"
Oct 19 08:24:55 nixus dockerd[1403]: time="2017-10-19T08:24:55.848698723-04:00" level=info msg="IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver
 2001:4860:4860::8844]"
Oct 19 08:24:56 nixus dockerd[1403]: /usr/local/bin/docker-entrypoint.sh: line 401: exec: --: invalid option
Oct 19 08:24:56 nixus dockerd[1403]: exec: usage: exec [-cl] [-a name] [command [arguments ...]] [redirection ...]
Oct 19 08:24:56 nixus dockerd[1403]: time="2017-10-19T08:24:56-04:00" level=error msg="setting up rule failed, [-t nat -A POSTROUTING -m ipvs --ipvs -d 10.0.0.0/24 -j SNAT --to-source 10.0.0.
3]:  (iptables failed: iptables --wait -t nat -A POSTROUTING -m ipvs --ipvs -d 10.0.0.0/24 -j SNAT --to-source 10.0.0.3: iptables: No chain/target/match by that name.
Oct 19 08:24:56 nixus dockerd[1403]:  (exit status 1))"
```

**Comment by NeQuissimus** [comment]
I don't know what to do about this, `iptables` is not something I really know how to mess with :)

**Comment by kuznero** [comment]
@NeQuissimus do you think restarting docker service can help after you added new kernel module?

**Comment by NeQuissimus** [comment]
I rebooted every time

**Comment by NeQuissimus** [comment]
Just found #24326. I will try that but am running a kernel build right now, so it will be a moment :)

**Comment by NeQuissimus** [comment]
I guess `virtualisation.docker.extraOptions = "--iptables=false";` does not work for Docker Swarm. (docker/swarm#1639)

It seems to be an `iptables` thing that does not work quite right, which I don't know much about.
Maybe @Mic92 can help, you seem to be knowledgeable when it comes to iptables :)
Do you think this is the issue mentioned in #24326 ?
How would we deal with it?

**Comment by kuznero** [comment]
Maybe somebody could elaborate on possible direction such that I can try to dig dipper myself? In the end of the day problem is caused by the combination of a new feature in docker 17.09 (that `rabbitmq:3.6-management` image is utilizing) and new features of NixOS 17.09. @NeQuissimus mentioned iptables, @Mic92 maybe you can also recommend somethings to narrow the direction a bit?

**Comment by kuznero** [comment]
I have learned that `sysctl -w net.ipv4.ip_forward=1` must be set. This might help to solve the issue. I will try this later today hopefully.

ping @NeQuissimus .

**Comment by NeQuissimus** [comment]
Interesting... I do have that value set to 1

**Comment by kuznero** [comment]
Indeed, checked that it does not have effect. Strangely enough, but my colleague observed it working briefly on `nixos-17.09` but after short time (10-15 minutes) it fallen back to the same weird behavior.

**Comment by NeQuissimus** [comment]
I still think that the weirdest thing here is that the test passes :D

**Comment by kuznero** [comment]
@NeQuissimus, can you please refer me to the tests you mentioned before?

**Comment by NeQuissimus** [comment]
https://gist.github.com/NeQuissimus/a902ce4192f24877b8c78b1bdde7e7c2

**Comment by kuznero** [comment]
@NeQuissimus the test is working in manual mode for me, when I pull official `rabbitmq:3.6.12-management` image. It works indeed. Though I could not run the full  automated test, it hangs after starting the `rabbit` service. I then changed `--detach=false` to `--detach` to make sure it proceeds. But that still hangs. Am I doing something wrong here.

**Comment by NeQuissimus** [comment]
Something in nixpkgs has changed since I last ran this... The test no longer pulls the Docker image for me...

**Comment by kuznero** [comment]
Try running it with my nixpkgs/pr/docker I linked above. I have reverted to using skopeo.

**Comment by kuznero** [comment]
@NeQuissimus, I just spotted another issue with the test - `docker service create` should use proper network, e.g. `docker service create --network ci`, which it does not currently. Will fix it and see if it still works.

**Comment by kuznero** [comment]
I ran manual tests with building up nix configuration in VirtualBox starting from almost blank one and to the configuration I typically use for development purposes. And I managed to figure out finally that this issue appears only when `networking.networkmanager.enable = true;`, otherwise, it works as it should! Default value for it is `false` that is why many simply don't notice it.

Here is a little description for the `networking.networkmanager.enable` option:

> Whether to use NetworkManager to obtain an IP address and other configuration for all network interfaces that are not manually configured. If enabled, a group networkmanager will be created. Add all users that should have permission to change network settings to this group.

So, solution `#1` is to simply disable network manager.

And, solution `#2` could be to make sure that user that runs `rabbitmq:3.6.12-management` is part of `networkmanager` group. But that would be strictly speaking a bit strange since e.g. user that runs `rabbitmq:3.6.12-management` has `uid=999` and it would be strange to make it part of `networkmanager` group. More strange it would be to have other such services configure in the same way. 

So, for now I am simply disabling `networkmanager`.
